### DISCUSSION TITLE
How to use the systemd directive "RestrictAddressFamilies" with "podman run"?

### DISCUSSION THREAD

**Discussion by eriksjolund** [containers/podman#14311] [Q&A] [answered]
Could the __systemd__ directive [__RestrictAddressFamilies__](https://www.freedesktop.org/software/systemd/man/systemd.exec.html#RestrictAddressFamilies=) be used to lock down `podman run` in a systemd user service?
The idea would be to get some extra protections against vulnerabilities in podman.

When I try __RestrictAddressFamilies=AF_UNIX__, `podman run` fails with `Error: OCI runtime error: crun: socket: Address family not supported by protocol`. (See below for the details)

 
Socket-activated sockets are excluded from the restrictions:

Quote from the [__systemd.exec__](https://www.freedesktop.org/software/systemd/man/systemd.exec.html#RestrictAddressFamilies=) man page
_Sockets passed into the process by other means (for example, by using socket activation with socket units, see systemd.socket(5)) are unaffected._

I did an experiment

```
mkdir -p  ~/.config/systemd/user
cd ~/.config/systemd/user
podman run --rm -d --name test docker.io/library/alpine sleep inf
podman generate systemd --name --new test | sed 's/\t/    /g' > test.service.orig
```

## Test 1:  _RestrictAddressFamilies=none_

The file _patch1_ contains

```diff
--- test.service.orig	2022-05-21 14:24:48.355738661 +0200
+++ test.service	2022-05-21 14:27:26.112727012 +0200
@@ -13,10 +13,15 @@
 Environment=PODMAN_SYSTEMD_UNIT=%n
 Restart=on-failure
 TimeoutStopSec=70
+RestrictAddressFamilies=none
 ExecStartPre=/bin/rm -f %t/%n.ctr-id
 ExecStart=/usr/bin/podman run \
     --cidfile=%t/%n.ctr-id \
     --cgroups=no-conmon \
+    --pull=never \
+    --log-driver=k8s-file \
+    --log-opt path=/tmp/test \
+    --network none \
     --rm \
     --sdnotify=conmon \
     --replace \
```

```
$ cp test.service.orig test.service
$ patch -p0 < patch1
patching file test.service
$ systemctl --user daemon-reload
$ systemctl --user start test.service
Job for test.service failed because the control process exited with error code.
See "systemctl --user status test.service" and "journalctl --user -xeu test.service" for details.
$ journalctl --user -xeu test.service | grep error| tail -1
May 21 14:45:53 asus podman[23080]: time="2022-05-21T14:45:53+02:00" level=error msg="Unable to write pod event: \"could not initialize socket to journald\""
$ 
```
__Result:__  `level=error msg="Unable to write pod event: \"could not initialize socket to journald\""`

## Test 2: *RestrictAddressFamilies=AF_UNIX*

Edit the file _test.service_ and replace

```
RestrictAddressFamilies=none
```

with

```
RestrictAddressFamilies=AF_UNIX
```

```
$ systemctl --user daemon-reload
$ systemctl --user reset-failed test.service
$ systemctl --user start test.service
Job for test.service failed because the control process exited with error code.
See "systemctl --user status test.service" and "journalctl --user -xeu test.service" for details.
$ journalctl --user -xeu test.service | grep Error | tail -1
May 21 14:52:09 asus podman[23374]: Error: OCI runtime error: crun: socket: Address family not supported by protocol
$ 
```
__Result:__  `Error: OCI runtime error: crun: socket: Address family not supported by protocol`

### Test 3:  *RestrictAddressFamilies=AF_UNIX AF_INET AF_INET6*

After also adding _AF_INET AF_INET6_ the service _test.service_ started successfully.

I'm not quite sure why the log says `image pull  docker.io/library/alpine` because `--pull=never` has been set.
```
$ journalctl --user -xeu test.service | tail -1
May 21 15:51:56 asus podman[23538]: 2022-05-21 15:51:55.973284311 +0200 CEST m=+0.025711802 image pull  docker.io/library/alpine
$ 
```

Is it possible to get it working with fewer privileges than:

```
RestrictAddressFamilies=AF_UNIX AF_INET AF_INET6
```
?

**Comment by mheon** [comment, marked as answer]
I can definitely say that `RestrictAddressFamilies=none` will not work - we use Unix sockets internally to handle container attach.

I am somewhat surprised that restricting to AF_UNIX only does not work - @giuseppe The error seems to be coming out of crun - thoughts?

@eriksjolund Does it work if you use runc instead?

**Reply by Luap99** [reply]
Greping the source code points to this: https://github.com/containers/crun/blob/e8b7c41ddb89503c05091c01085b49bea5784e85/src/libcrun/linux.c#L4701

I think it would be possible to drop this and use netlink instead, but I think in this case you would need to allow AF_NETLINK sockets.

**Reply by rhatdan** [reply]
@giuseppe WDYT?

**Reply by eriksjolund** [reply]
> @eriksjolund Does it work if you use runc instead?

@mheon  `runc` works with __AF_UNIX AF_NETLINK__

| | crun | runc |
| -- | -- | -- |
| AF_UNIX |  ❌ | ❌ |
| AF_NETLINK |  ❌ | ❌ |
| AF_UNIX AF_INET AF_INET6 |  ✅ | ❌ |
| AF_UNIX AF_NETLINK |  ❌ | ✅ |

**Reply by giuseppe** [reply]
I think it should be doable to add a fallback to netlink when ` (AF_INET, SOCK_DGRAM, 0)` fails, so we support both `AF_UNIX AF_INET AF_INET6` and `AF_UNIX AF_NETLINK` combinations

**Reply by giuseppe** [reply]
interested to open a PR? :-)

**Reply by eriksjolund** [reply]
Not at the moment. (Currently I'm not familiar with AF_NETLINK)

**Reply by eriksjolund** [reply]
I created a feature request

* https://github.com/containers/crun/issues/929


A sidenote: I did a successful test with __runc__ and the echo server
__ghcr.io/eriksjolund/socket-activate-echo__ with one socket-activated TCP socket
and `RestrictAddressFamilies=AF_UNIX AF_NETLINK`.


I noticed there is problem when using 3 socket-activated sockets with __runc__
* https://github.com/opencontainers/runc/issues/3488

**Reply by eriksjolund** [reply]
I verified that the systemd restriction works. As expected, the combination

* runc
* `RestrictAddressFamilies=AF_UNIX AF_NETLINK`
* `--pull=always`
* `--network=none`
* `--log-driver=k8s-file`
* `--log-opt path=/tmp/test.log`
* one socket-activated TCP socket
    ```
    ● test.socket - test
         Loaded: loaded (/home/test138/.config/systemd/user/test.socket; disabled; vendor preset: disabled)
         Active: active (listening) since Thu 2022-05-26 10:13:00 CEST; 3s ago
          Until: Thu 2022-05-26 10:13:00 CEST; 3s ago
       Triggers: ● test.service
         Listen: 127.0.0.1:9000 (Stream)
         CGroup: /user.slice/user-1047.slice/user@1047.service/app.slice/test.socket
    
    May 26 10:13:00 asus systemd[10686]: Listening on test.socket - test.
    ```

fails as it should because Podman is not able to pull the container image.
`journalctl` shows

```
$ journalctl --user -xe -u test.service | grep -i error | tail -2
May 26 10:09:54 asus podman[28272]: Error: initializing source docker://ghcr.io/eriksjolund/socket-activate-echo:latest: pinging container registry ghcr.io: Get "https://ghcr.io/v2/": dial tcp 140.82.121.34:443: socket: address family not supported by protocol
May 26 10:09:54 asus podman[28291]: Error: error reading CIDFile: open /run/user/1047/test.service.ctr-id: no such file or directory
$ 
```

**Reply by eriksjolund** [reply]
_If running in user mode, or in system mode, but without the CAP_SYS_ADMIN capability (e.g. setting User=), NoNewPrivileges=yes is implied._
quote from the systemd documentation [__RestrictAddressFamilies__](https://www.freedesktop.org/software/systemd/man/systemd.exec.html#RestrictAddressFamilies=) 

I noticed that the Podman user namespace needs to be created explicitly in another user service  because `NoNewPrivileges=yes` has the effect that __newuidmap__  runs without its extra capabilities and thus fails. See the discussion in
*  https://github.com/containers/podman/discussions/14404

**Comment by vrothberg** [comment]
For completeness, I want to add that customizing the generated units files is not supported.  While we will do our best to resolve issues as they come in, we can only support that Podman generates.
